Support reverse lookups from the host file.

cmcfarlen opened the pull request:
Refactor host file code to make it more testable and unify lookups
Add unit test for HostFile

cmcfarlen commented:
Fixes #9025

cmcfarlen commented:
While this supports reverse lookups from the hosts file, there is some issue with deterministic selection of the name that is resolved for a given IP since the order depends on the appearance of the file as well as how we parse the hosts file.  Currently the unit test is broken because of this nondeterministic behavior.  Should we try to guarantee a deterministic resolution or should we just return any host with the assigned IP?  I'm not sure how ATS handles multiple hosts returned from a reverse lookup currently.

Should this "reverse lookup from the hosts file" be a configurable feature?

SolidWallOfCode commented:
Is it indeterminate because an unordered map is used? I think it would be fine to say which ever name is first associated with an IP address is the reverse lookup name.

SolidWallOfCode commented on iocore/hostdb/HostFile.cc:
```diff
@@ -0,0 +1,141 @@
+/* Container for temporarily holding data from the host file. For each FQDN there is a vector of IPv4
+ * and IPv6 addresses. These are used to generate the HostDBRecord instances that are stored persistently.
+ */
+using HostAddrMap = std::unordered_map<std::string_view, std::tuple<std::vector<IpAddr>, std::vector<IpAddr>>>;
```
If you're going to split it anyway, you might as well use `Ip4Addr` and `Ip6Addr`.

SolidWallOfCode commented:
I wonder if `Lexicon` would work in this situation. It supports primary and secondary names vs. a numeric value, which seems like the situation here.

bryancall commented:
[approve ci autest]

maskit reviewed (changes requested):
Is this for 10-Dev or not? It's marked as 10-Dev but the target branch is master.

cmcfarlen commented:
> Is this for 10-Dev or not? It's marked as 10-Dev but the target branch is master.

I moved it to 10.0 milestone but idk how to remove it from the 10-dev project.

bneradt commented:
> > Is this for 10-Dev or not? It's marked as 10-Dev but the target branch is master.
> 
> I moved it to 10.0 milestone but idk how to remove it from the 10-dev project.

I removed it. I just uncheck-marked the 10-Dev project.

cmcfarlen commented:
> > > Is this for 10-Dev or not? It's marked as 10-Dev but the target branch is master.
> > 
> > 
> > I moved it to 10.0 milestone but idk how to remove it from the 10-dev project.
> 
> I removed it. I just uncheck-marked the 10-Dev project.

Thanks, I must not have that permission as I didn't see a checkmark to uncheck!
